### ISSUE TITLE
Celery keep executing the same task despite SUCCESS state

### ISSUE THREAD

**Issue by krnr** [celery/celery#6871]
# Checklist

- [ ] I have verified that the issue exists against the `master` branch of Celery.
- [ ] This has already been asked to the [discussion group](https://groups.google.com/forum/#!forum/celery-users) first.
- [X] I have read the relevant section in the
  [contribution guide](http://docs.celeryproject.org/en/latest/contributing.html#other-bugs)
  on reporting bugs.
- [X] I have checked the [issues list](https://github.com/celery/celery/issues?q=is%3Aissue+label%3A%22Issue+Type%3A+Bug+Report%22+-label%3A%22Category%3A+Documentation%22)
  for similar or identical bug reports.
- [X] I have checked the [pull requests list](https://github.com/celery/celery/pulls?q=is%3Apr+label%3A%22PR+Type%3A+Bugfix%22+-label%3A%22Category%3A+Documentation%22)
  for existing proposed fixes.
- [ ] I have checked the [commit log](https://github.com/celery/celery/commits/master)
  to find out if the bug was already fixed in the master branch.
- [ ] I have included all related issues and possible duplicate issues
  in this issue (If there are none, check this box anyway).

## Mandatory Debugging Information

- [X] I have included the output of ``celery -A proj report`` in the issue.
    (if you are not able to do this, then at least specify the Celery
     version affected).
- [ ] I have verified that the issue exists against the `master` branch of Celery.
- [X] I have included the contents of ``pip freeze`` in the issue.
- [ ] I have included all the versions of all the external dependencies required
  to reproduce this bug.

## Optional Debugging Information

- [ ] I have tried reproducing the issue on more than one Python version
  and/or implementation.
- [ ] I have tried reproducing the issue on more than one message broker and/or
  result backend.
- [ ] I have tried reproducing the issue on more than one version of the message
  broker and/or result backend.
- [ ] I have tried reproducing the issue on more than one operating system.
- [ ] I have tried reproducing the issue on more than one workers pool.
- [ ] I have tried reproducing the issue with autoscaling, retries,
  ETA/Countdown & rate limits disabled.
- [ ] I have tried reproducing the issue after downgrading
  and/or upgrading Celery and its dependencies.

#### Related Issues

- https://github.com/celery/celery/issues/5935

#### Possible Duplicates

- None

## Environment & Settings
**Celery version**: 4.4.3
<details>
<summary><b><code>celery report</code> Output:</b></summary>
<p>

```
software -> celery:4.4.3 (cliffs) kombu:4.6.10 py:3.8.0
            billiard:3.6.4.0 redis:3.5.3
platform -> system:Linux arch:64bit, ELF
            kernel version:5.3.0-40-generic imp:CPython
loader   -> celery.loaders.app.AppLoader
settings -> transport:redis results:redis://worker-wm-01.local:6379/

broker_url: 'redis://worker-wm-01.local:6379//'
result_backend: 'redis://worker-wm-01.local:6379/'
include: ['tasks']
broker: 'redis://worker-wm-01.local:6379'
broker_transport_options: {
 'interval_max': 5, 'interval_start': 0, 'interval_step': 0.4, 'max_retries': 5}
broker_pool_limit: None
backend: 'redis://worker-wm-01.local:6379/'
redis_backend_health_check_interval: 120
task_acks_late: True
worker_prefetch_multiplier: 1
```

</p>
</details>

# Steps to Reproduce

## Required Dependencies

* **Minimal Python Version**: 3.8.0
* **Minimal Celery Version**: 4.4.3
* **Minimal Kombu Version**: 4.6.10
* **Minimal Broker Version**: N/A or Unknown
* **Minimal Result Backend Version**: N/A or Unknown
* **Minimal OS and/or Kernel Version**: N/A or Unknown
* **Minimal Broker Client Version**: N/A or Unknown
* **Minimal Result Backend Client Version**: N/A or Unknown

### Python Packages

<details>
<summary><b><code>pip freeze</code> Output:</b></summary>
<p>

```
absl-py==0.12.0         
alembic==1.4.1   
amqp==2.6.1 
ase==3.21.1       
astroid==2.5.4   
attrs==20.3.0    
billiard==3.6.4.0
cachetools==4.2.1 
celery==4.4.3                
certifi==2020.12.5  
chardet==3.0.4
click==7.1.2 
cloudpickle==1.6.0  
coverage==5.5         
cycler==0.10.0      
databricks-cli==0.14.3
dataclasses==0.6        
decorator==4.4.2
docker==5.0.0             
entrypoints==0.3                 
flake8==3.9.1   
Flask==1.1.2
Flask-PyMongo==2.3.0    
future==0.18.2 
gitdb==4.0.7   
GitPython==3.1.14
google-auth==1.29.0
google-auth-oauthlib==0.4.4
googledrivedownloader==0.4
greenlet==1.0.0       
grpcio==1.37.0       
gunicorn==20.1.0    
h11==0.8.1          
h2==3.2.0               
h5py==3.2.1 
hpack==3.0.0 
http3==0.6.7             
hyperframe==5.2.0
idna==2.10  
iniconfig==1.1.1
isodate==0.6.0       
isort==5.8.0            
itsdangerous==1.1.0
Jinja2==2.11.3
joblib==1.0.1       
kiwisolver==1.3.1
kombu==4.6.10  
lazy-object-proxy==1.6.0
llvmlite==0.36.0 
Mako==1.1.4 
Markdown==3.3.4   
MarkupSafe==1.1.1
matplotlib==3.4.1
mccabe==0.6.1  
minio==7.0.3      
mlflow==1.15.0               
networkx==2.5.1     
numba==0.53.1
numpy==1.20.2
oauthlib==3.1.0     
packaging==20.9       
pandas==1.2.4       
pefile==2019.4.18  
Pillow==8.2.0           
pluggy==0.13.1
prometheus-client==0.10.1 
prometheus-flask-exporter==0.18.1
protobuf==3.15.8
py==1.10.0 
pyasn1==0.4.8           
pyasn1-modules==0.2.8
pycodestyle==2.7.0
pyflakes==2.3.1
pylint==2.7.4
pymongo==3.11.3
pyparsing==2.4.7
pytest==6.2.3
python-dateutil==2.8.1
python-dotenv==0.17.0
python-editor==1.0.4
python-louvain==0.15
pytorch-lightning==0.9.0
pytz==2021.1
PyYAML==5.4.1
querystring-parser==1.2.4
rdflib==5.0.0
redis==3.5.3
requests==2.25.1
requests-async==0.6.2
requests-oauthlib==1.3.0
rfc3986==1.4.0
rsa==4.7.2
scikit-learn==0.24.1
scipy==1.6.2
seaborn==0.11.1
six==1.15.0
smart-open==5.0.0
smmap==4.0.0
SQLAlchemy==1.4.11
sqlparse==0.4.1
structlog==21.1.0
tabulate==0.8.9
tensorboard==2.2.0
tensorboard-plugin-wit==1.8.0
threadpoolctl==2.1.0
toml==0.10.2
torch==1.7.0
torch-cluster==1.5.8
torch-geometric==1.6.3
torch-scatter==2.0.5
torch-sparse==0.6.8
torch-spline-conv==1.2.0
tqdm==4.60.0
typing-extensions==3.7.4.3
urllib3==1.26.4
uWSGI==2.0.19.1
vine==1.3.0
websocket-client==0.58.0
Werkzeug==1.0.1
wrapt==1.12.1

```

</p>
</details>

### Other Dependencies

<details>
<p>
N/A
</p>
</details>

## Minimally Reproducible Test Case

<details>
<p>

```python
@shared_task(bind=True)
def process(self, filename):
    run_long_process()  # more than 2 hours
    return filename

celery_app.send_task("tasks.process", args=(filename,), queue=queue)

```

</p>
</details>

# Expected Behavior
Task is executed only once.

# Actual Behavior

the task is keep executing after having got state SUCCESS once:

```
>>> from app import celery_app
>>> ar = celery_app.AsyncResult('46906783-4351-4397-8893-dfc69c8421cf')
>>> ar.state
'SUCCESS'
```

```
[2021-07-22 21:03:05,141: INFO/MainProcess] Received task: tasks.process[46906783-4351-4397-8893-dfc69c8421cf]  
[2021-07-22 23:22:02,694: INFO/MainProcess] Received task: tasks.process[46906783-4351-4397-8893-dfc69c8421cf]  
[2021-07-23 01:41:35,151: INFO/MainProcess] Received task: tasks.process[46906783-4351-4397-8893-dfc69c8421cf]  
[2021-07-23 04:02:10,839: INFO/MainProcess] Received task: tasks.process[46906783-4351-4397-8893-dfc69c8421cf]  
```

I've noticed that unlike the similar issue, the default value of 1 hour visibility timeout doesn't affect me. Tasks between 1 and 2 hours long are performed only once, but this one takes about 140 minutes and keeps reappearing. I also read that actually visibility timeout doesn't affect anything and this kind of bugs happens mostly with Redis backend. Like in other cases, change to Rabbit is not an option.

I would like to discover more about this issue, but i'm not sure where to head at. and whether it is similar to other cases or a new one. and what exactly is the nature of getting tasks to a worker (is it only that [this method](https://github.com/celery/celery/blob/f462a437e3371acb867e94b52c2595b6d0a742d8/celery/worker/request.py#L587) must be called for a task to never be picked again or it doesn't matter in some cases)?

**Comment by jonaslimads** [comment]
This same bug happened with us after upgrading RabbitMQ from 3.8.16 to [3.9.0](https://github.com/rabbitmq/rabbitmq-server/releases/tag/v3.9.0). RabbitMQ 3.8.16 worked just fine, but 3.9.0 has introduced this bug.

We run the old Celery 4.4.0, but it seems the longer it takes to upgrade Celery, the more sensitive is it to changes in RabbitMQ, so we better upgrade to the current version of Celery and Kombu and other dependencies.

**Comment by jspaans91** [comment]
It's looks like the same behavior we experience after upgrading RabbitMQ to version 3.9.0. Our tasks have duration of milliseconds for sending mail by API, but after the upgrade the message is send 5 times instead of 1.

In our Celery log we saw also the following errors during the handling of the tasks:

`[2021-07-28 20:45:32,564: INFO/MainProcess] Connected to amqp://ehvappprd:**@127.0.0.1:5672/ehvappprd
[2021-07-28 20:45:32,573: INFO/MainProcess] mingle: searching for neighbors
[2021-07-28 20:45:33,588: INFO/MainProcess] mingle: all alone
[2021-07-28 20:45:33,602: INFO/MainProcess] Received task: logistiek.tasks.send_email_with_order[f962f1aa-d05c-4668-aa2f-e01603af2f38]  
[2021-07-28 20:45:33,602: INFO/MainProcess] Received task: logistiek.tasks.send_email_with_order[c2d60c02-faf9-4c07-a52e-0b1c7d745162]  
[2021-07-28 20:45:33,754: WARNING/MainProcess] consumer: Connection to broker lost. Trying to re-establish the connection...
Traceback (most recent call last):
  File "/apps/ehvappprd/venv/lib/python3.8/site-packages/celery/worker/consumer/consumer.py", line 311, in start
    blueprint.start(self)
  File "/apps/ehvappprd/venv/lib/python3.8/site-packages/celery/bootsteps.py", line 116, in start
    step.start(parent)
  File "/apps/ehvappprd/venv/lib/python3.8/site-packages/celery/worker/consumer/consumer.py", line 592, in start
    c.loop(*c.loop_args())
  File "/apps/ehvappprd/venv/lib/python3.8/site-packages/celery/worker/loops.py", line 81, in asynloop
    next(loop)
  File "/apps/ehvappprd/venv/lib/python3.8/site-packages/kombu/asynchronous/hub.py", line 361, in create_loop
    cb(*cbargs)
  File "/apps/ehvappprd/venv/lib/python3.8/site-packages/kombu/transport/base.py", line 235, in on_readable
    reader(loop)
  File "/apps/ehvappprd/venv/lib/python3.8/site-packages/kombu/transport/base.py", line 217, in _read
    drain_events(timeout=0)
  File "/apps/ehvappprd/venv/lib/python3.8/site-packages/amqp/connection.py", line 523, in drain_events
    while not self.blocking_read(timeout):
  File "/apps/ehvappprd/venv/lib/python3.8/site-packages/amqp/connection.py", line 529, in blocking_read
    return self.on_inbound_frame(frame)
  File "/apps/ehvappprd/venv/lib/python3.8/site-packages/amqp/method_framing.py", line 53, in on_frame
    callback(channel, method_sig, buf, None)
  File "/apps/ehvappprd/venv/lib/python3.8/site-packages/amqp/connection.py", line 535, in on_inbound_method
    return self.channels[channel_id].dispatch_method(
  File "/apps/ehvappprd/venv/lib/python3.8/site-packages/amqp/abstract_channel.py", line 143, in dispatch_method
    listener(*args)
  File "/apps/ehvappprd/venv/lib/python3.8/site-packages/amqp/connection.py", line 665, in _on_close
    raise error_for_code(reply_code, reply_text,
amqp.exceptions.InternalError: (0, 0): (541) INTERNAL_ERROR
[2021-07-28 20:45:33,763: INFO/MainProcess] Connected to amqp://ehvappprd:**@127.0.0.1:5672/ehvappprd
[2021-07-28 20:45:33,772: INFO/MainProcess] mingle: searching for neighbors
[2021-07-28 20:45:34,788: INFO/MainProcess] mingle: all alone
[2021-07-28 20:45:34,802: INFO/MainProcess] Received task: logistiek.tasks.send_email_with_order[c2d60c02-faf9-4c07-a52e-0b1c7d745162]  
[2021-07-28 20:45:34,904: CRITICAL/MainProcess] Couldn't ack 2, reason:BrokenPipeError(32, 'Broken pipe')
Traceback (most recent call last):
  File "/apps/ehvappprd/venv/lib/python3.8/site-packages/kombu/message.py", line 128, in ack_log_error
    self.ack(multiple=multiple)
  File "/apps/ehvappprd/venv/lib/python3.8/site-packages/kombu/message.py", line 123, in ack
    self.channel.basic_ack(self.delivery_tag, multiple=multiple)
  File "/apps/ehvappprd/venv/lib/python3.8/site-packages/amqp/channel.py", line 1391, in basic_ack
    return self.send_method(
  File "/apps/ehvappprd/venv/lib/python3.8/site-packages/amqp/abstract_channel.py", line 57, in send_method
    conn.frame_writer(1, self.channel_id, sig, args, content)
  File "/apps/ehvappprd/venv/lib/python3.8/site-packages/amqp/method_framing.py", line 183, in write_frame
    write(view[:offset])
  File "/apps/ehvappprd/venv/lib/python3.8/site-packages/amqp/transport.py", line 352, in write
    self._write(s)
BrokenPipeError: [Errno 32] Broken pipe`

After downgrading to version 3.8.19 of RabbitMQ problems were solved.
We've version 5.1.2 of Celery on Ubuntu 20.04.2 LTS.

**Comment by krnr** [comment]
but we use redis as a backend, so I don't think it's "broker-related"

**Comment by thedrow** [comment]
@krnr Valid point, but this could point us in the right direction.
If we can tell which area of the code causes the problem, that would lead to a solution.

Parhaps we should try to reproduce the issue with RabbitMQ 3.9 if this reproduces more reliably when Celery uses it as a message broker?

**Comment by thedrow** [comment]
Since no other input was provided, I'm closing this as a duplicate of #5935.

You can use the [`worker_deduplicate_successful_tasks`](https://docs.celeryproject.org/en/stable/userguide/configuration.html#worker-deduplicate-successful-tasks) setting to mitigate this issue.
